### PR TITLE
[P4Testgen] Create abstract execution state with utility functions - move copy-in/out to control/parser blocks.

### PR DESCRIPTION
**Pull request by fruffy**
Other P4Tools modules can also benefit from execution state, but Testgen's execution state was too specialized.  This PR cleans up the execution state class and introduces an abstract, general execution state. Coupled with that, we introduce several utilities that are beneficial for performance and simplification of core functions. Among others, we simplify the way we handle copy-in/out and struct assignments. 
There were also other small improvements:

- Added some more table utilities.
- Reduce some unnecessary execution state copies.
- Do not offload parser state resolution to the IR::PathExpression visitor. Instead, resolve select expressions and parser states directly.

Performance for generating 10000 tests for `pins_middleblock.p4`.
Before:
```
Executed in  495.09 secs    fish           external
   usr time  483.90 secs  409.00 micros  483.90 secs
   sys time    8.98 secs  252.00 micros    8.98 secs
```

After:
```
Executed in  431.90 secs    fish           external
   usr time  422.09 secs  519.00 micros  422.09 secs
   sys time    7.86 secs  320.00 micros    7.86 secs

```
A ~13% reduction.

### CODE REVIEW

File: backends/p4tools/common/lib/variables.cpp [hunk 1]
```diff
@@ -53,4 +52,14 @@ const IR::TaintExpression *ToolsVariables::getTaintExpression(const IR::Type *ty
...
 }
 
-}  // namespace P4Tools
+IR::StateVariable convertReference(const IR::Expression *ref) {
+    if (const auto *member = ref->to<IR::Member>()) {
+        return member;
+    }
+    // Local variable.
+    const auto *path = ref->checkedTo<IR::PathExpression>();
+    // return getStateVariable(path->type, path->path->name);
```
Comment (pkotikal) [inline backends/p4tools/common/lib/variables.cpp:61]: Remove comment.

File: backends/p4tools/modules/testgen/core/small_step/expr_stepper.cpp [hunk 2]
```diff
@@ -137,10 +136,9 @@ bool ExprStepper::preorder(const IR::MethodCallExpression *call) {
             BUG_CHECK(method->expr, "Method call has unexpected format: %1%", call);
 
             // Handle table calls.
-            if (const auto *table = state.getTableType(method)) {
-                auto &nextState = state.clone();
-                nextState.replaceTopBody(Continuation::Return(table));
-                result->emplace_back(nextState);
+            if (const auto *table = state.findTable(method)) {
```
Comment (pkotikal) [inline backends/p4tools/modules/testgen/core/small_step/expr_stepper.cpp:139]: Check for null table?

**Review by pkotikal — APPROVED** [review]
Looks good, thanks!

File: backends/p4tools/common/core/abstract_execution_state.cpp [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+/* =============================================================================================
+ *  Accessors
+ * ============================================================================================= */
+
+bool AbstractExecutionState::exists(const IR::StateVariable &var) const { return env.exists(var); }
+
+const SymbolicEnv &AbstractExecutionState::getSymbolicEnv() const { return env; }
+
+void AbstractExecutionState::printSymbolicEnv(std::ostream &out) const {
+    // TODO(fruffy): How do we do logging here?
```
[the remark this answers is not in the document]
Comment (fruffy) [reply backends/p4tools/common/core/abstract_execution_state.cpp:28]: Yes, that would be great. For now, printing to stdout here has been convenient.

File: backends/p4tools/modules/testgen/core/program_info.h [hunk 4]
```diff
@@ -35,6 +35,9 @@ class ProgramInfo : public ICastable {
     /// The constraints imposed by the target.
     std::optional<const IR::Expression *> targetConstraints = std::nullopt;
 
+    /// Maps the programmable blocks in the P4 program to their canonical counterpart.
+    std::map<cstring, cstring> blockMap;
```
[the remark this answers is not in the document]
Comment (fruffy) [reply backends/p4tools/modules/testgen/core/program_info.h:39]: There is no need for ordering, but it can't hurt. Changed it.

**Pull request merged**